File: c++/SingledLinkedList.cpp
#include <iostream>

using namespace std;

class Node {
private:
  int data;
  Node *next;
public:
  Node(int value) 
  { 
    data = value;
    next = NULL;
  }

  int getData() {
    return data;
  }

  Node *getNext() {
    return next;
  }

  void setNext(Node *node) {
    next = node;
  }
};

class LinkedList {
private:
  Node *head;
  Node *tail;
public:

  LinkedList() {
    head = NULL;
    tail = NULL;
  }

  void addToTail(int value) {
    Node *node = new Node(value);
    if (head == NULL) {
      head = node;
      tail = node;
    }
    else {
      tail->setNext(node);
      tail = node;
    }
  }
  
  void removeLastNode() {
    //find the prev of tail
    Node *prevNode = NULL;
    Node *curNode = head;
    while (curNode != tail) {
      prevNode = curNode;
      curNode = curNode->getNext();
    }
    
    tail = prevNode;
    tail->setNext(NULL);
  }

  void traverse() {
    Node *curNode = head;
    while (curNode != NULL) {
      cout << curNode->getData() << "\n";
      curNode = curNode->getNext();
    }
  }



};

int main() {

  LinkedList *list = new LinkedList();
  list->addToTail(1);
  list->addToTail(3);
  list->addToTail(2);
  list->traverse();
  list->removeLastNode();
  list->traverse();
  
}

File: php/SingledLinkedList.php
<?php

  // php code is wrapped in <?php tags

  class Node {
    public $data;
    public $next;
    function __construct($data){
      $this->data = $data;
      $this->next = null;
    }
  }

  class LinkedList {
    public $head;
    public $tail;
    
    function __construct(){
      $this->head = null;
      $this->tail = null;
    }
    
    function addToTail($data) {
      $node = new Node($data);
      if ($this->head == null) {
        $this->head = $node;
        $this->tail = $node;
      } else {
        $this->tail->next = $node;
        $this->tail = $node;
      }
    }
    
    function removeLastNode() {
      $curNode = $this->head;
      while($curNode != $this->tail) {
        $prevNode = $curNode;
        $curNode = $curNode->next;
      }
      
      $this->tail = $prevNode;
      $this->tail->next = null;
      
    }
    
    function traverse() {
      $curNode = $this->head;
      while($curNode != null) {
        echo $curNode->data."\n";
        $curNode = $curNode->next;
      }
    }
  }

  $list = new LinkedList();
  $list->addToTail("abc");
  $list->addToTail("def");
  $list->traverse();
  echo "===\n";
  $list->removeLastNode();
  $list->traverse();
  
  /*for ( $i = 0; $i < 5; $i++) {
    echo "Hello, World\n";
  }*/
?>
